src/lib.rs

```diff
@@ -1,27 +1,50 @@
-use std::collections::HashMap;
-
 use borsh::schema::{BorshSchemaContainer, Declaration, Definition, Fields, VariantName};
 use schemars::schema::{RootSchema, Schema};
-use serde::{Deserialize, Serialize};
+use semver::Version;
+use serde::{de, Deserialize, Deserializer, Serialize};
+use std::collections::HashMap;
 
 #[doc(hidden)]
 #[cfg(feature = "__chunked-entries")]
 #[path = "private.rs"]
 pub mod __private;
 
 /// Current version of the ABI schema format.
-pub const SCHEMA_VERSION: &str = env!("CARGO_PKG_VERSION");
+pub const SCHEMA_VERSION_MAJOR: u64 = 0;
+pub const SCHEMA_VERSION_MINOR: u64 = 1;
+pub const SCHEMA_VERSION: &str = "0.1.0";
 
 /// Contract ABI.
 #[derive(Clone, Serialize, Deserialize, Debug, PartialEq)]
 pub struct AbiRoot {
     /// Semver of the ABI schema format.
-    #[serde(rename = "abi_schema_version")]
+    #[serde(deserialize_with = "ensure_current_version")]
     pub schema_version: String,
     /// Metadata information about the contract.
     pub metadata: AbiMetadata,
     /// Core ABI information (functions and types).
-    pub abi: AbiEntry,
+    pub body: AbiBody,
+}
+
+fn ensure_current_version<'de, D: Deserializer<'de>>(d: D) -> Result<String, D::Error> {
+    let unchecked = String::deserialize(d)?;
+    let version = Version::parse(&unchecked)
+        .map_err(|_| de::Error::custom("expected `schema_version` to be a valid semver value"))?;
+    if version.major != SCHEMA_VERSION_MAJOR || version.minor != SCHEMA_VERSION_MINOR {
+        let expected_version = Version::parse(SCHEMA_VERSION).unwrap();
+        if version < expected_version {
+            return Err(de::Error::custom(format!(
+                "expected `schema_version` to be ~{}.{}, but got {}: consider re-generating your ABI file with a newer version of SDK and cargo-near",
```

Comment:
This seems strange that we would restrict to using this lib with only one specific version. This leads that whenever this version is changed, any ABIs used currently will be unusable with any of our tooling. Isn't the point that we want this to be backwards compatible as much as possible? There might be some cases where we might want to rename/restructure/reduce the format without adding any more data and this can be done in a backwards compatible way.

Want to avoid breaking usage such as `someone embeds ABI in contract -> we update format -> this breaks the usage of that ABI for use in any tooling (and would force someone to re-deploy if they want the same pattern)`

Reply:
Let me explain my reasoning here. While we are in early-semver we are probably going to break compatibility in minor versions. I can modify this PR to allow a range of minor versions (e.g. >= 0.2 && < 0.4 if we know that 0.3 changes are purely optionally additive), but seemed redundant to do right now before the need arose. Once we reach 1.0, we are going to switch to only checking the major version and forget about the minor one.

Do you think there any viable alternatives? Creating a separate model per `major.minor` combination seems like maintenance hell to me, but we could try that.